subsys/power/device_pm.c

```diff
@@ -0,0 +1,301 @@
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	request = dev->pm->async_request;
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_INPROG;
+	k_sem_give(&dev->pm->lock);
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n", request, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+	/* Set async status to done */
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_DONE;
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_state_handler(struct device *dev,
+				   u32_t target_state, u32_t pm_flags)
+{
+	u32_t current_state;
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto handler_out;
+	}
+
+	/* Check if current power state is same as target
+	 * state so that we can skip processing new request.
+	 */
+	if (current_state == target_state) {
+		if (pm_flags & DEVICE_PM_ASYNC) {
+			if (dev->pm->async_request == target_state) {
+				goto handler_out;
+			}
+		} else {
+			goto handler_out;
+		}
+	}
+
+	/* Schedule the worker for new async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->async_request = target_state;
+		dev->pm->async_req_status = DEVPM_ASYNC_REQ_QUEUED;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+
+		return DEVICE_PM_RET_QUEUED;
+	}
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, target_state);
+	if (ret < 0) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+						target_state, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+handler_out:
+	return ret;
+}
+
+static int device_pm_handle_get(struct device *dev, u32_t pm_flags)
+{
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		ret = -ERANGE;
+		goto get_out;
+	}
+	dev->pm->usage++;
+
+	if (!dev->pm->enable) {
+		goto get_out;
+	}
+
+	if (dev->pm->usage > 0) {
+		ret = device_pm_state_handler(dev,
```

Comment:
If the call is synchronous, it block all asynchronous API due to `dev->pm->lock`.

Reply:
Ok will see if i can improve the logic.